Add recipe for elfeed-tube

karthink opened the pull request:
### Brief summary of what the package does

Elfeed Tube is an extension for Elfeed, the feed reader for Emacs, that enhances Youtube RSS feed subscriptions.

Typically Youtube RSS feeds contain only the title and author of each video. Elfeed Tube adds video descriptions, thumbnails, durations, chapters and "live" transcrips to video entries. See https://github.com/karthink/elfeed-tube for demos. This information can optionally be added to the entry in the Elfeed database.

### Direct link to the package repository

https://github.com/karthink/elfeed-tube

### Your association with the package

I'm the author and maintainer.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

<!-- Please confirm with `x`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [ ] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->
Checkdoc insists on docstrings for all functions, including one-line `defsubst`s. I've decided to ignore this. All interactive commands provided by the package are documented.

riscy commented:
> Checkdoc insists on docstrings for all functions, including one-line defsubsts. I've decided to ignore this. All interactive commands provided by the package are documented.

I realize it's pedantic and probably more than I would personally normally do, but at the moment it's on the checklist and asked of all authors equally.  I'll include those here along with some other issues:

`elfeed-tube-utils.el` with byte-compile using Emacs 28.1:
```
elfeed-tube-utils.el:87:18: Warning: reference to free variable
    `elfeed-tube-youtube-regexp'
In elfeed-tube-add--display-channels:
elfeed-tube-utils.el:222:22: Warning: reference to free variable
    `elfeed-tube-youtube-regexp'
In elfeed-tube--aio-fetch:
elfeed-tube-utils.el:330:36: Warning: reference to free variable
    `elfeed-tube--max-retries'
In elfeed-tube--video-p:
elfeed-tube-utils.el:353:11: Warning: defsubst `elfeed-tube--video-p' was used
    before it was defined
elfeed-tube-utils.el:356:5: Warning: reference to free variable
    `elfeed-tube-youtube-regexp'
In elfeed-tube--playlist-p:
elfeed-tube-utils.el:361:11: Warning: defsubst `elfeed-tube--playlist-p' was
    used before it was defined
elfeed-tube-utils.el:364:5: Warning: reference to free variable
    `elfeed-tube-youtube-regexp'
In elfeed-tube--channel-p:
elfeed-tube-utils.el:369:11: Warning: defsubst `elfeed-tube--channel-p' was
    used before it was defined
elfeed-tube-utils.el:372:5: Warning: reference to free variable
    `elfeed-tube-youtube-regexp'
In elfeed-tube--fake-entry:
elfeed-tube-utils.el:392:40: Warning: reference to free variable
    `elfeed-tube--api-videos-path'
elfeed-tube-utils.el:378:25: Warning: reference to free variable
    `elfeed-tube-youtube-regexp'
In end of data:
elfeed-tube-utils.el:454:56: Warning: the function `prop-match-end' is not
    known to be defined.
elfeed-tube-utils.el:454:27: Warning: the function `prop-match-beginning' is
    not known to be defined.
elfeed-tube-utils.el:446:30: Warning: the function
    `text-property-search-backward' is not known to be defined.
elfeed-tube-utils.el:445:32: Warning: the function
    `text-property-search-forward' is not known to be defined.
elfeed-tube-utils.el:404:22: Warning: the function `elfeed-entry--create' is
    not known to be defined.
elfeed-tube-utils.el:415:24: Warning: the function `elfeed-feed--create' is
    not known to be defined.
elfeed-tube-utils.el:420:23: Warning: the function `elfeed-tube--fetch-1' is
    not known to be defined.
elfeed-tube-utils.el:426:28: Warning: the function `elfeed-tube-show' is not
    known to be defined.
elfeed-tube-utils.el:423:23: Warning: the function `elfeed-entry-title' is not
    known to be defined.
elfeed-tube-utils.el:421:34: Warning: the function `elfeed-show-entry' is not
    known to be defined.
elfeed-tube-utils.el:333:27: Warning: the function `elfeed-tube-curl-enqueue'
    is not known to be defined.
elfeed-tube-utils.el:348:31: Warning: the function `elfeed-tube--attempt-log'
    is not known to be defined.
elfeed-tube-utils.el:347:14: Warning: the function `elfeed-tube-log' is not
    known to be defined.
elfeed-tube-utils.el:286:20: Warning: the function `elfeed-add-feed' is not
    known to be defined.
elfeed-tube-utils.el:130:45: Warning: the function
    `elfeed-tube--get-invidious-url' is not known to be defined.
elfeed-tube-utils.el:138:40: Warning: the function
    `elfeed-tube--nrotate-invidious-servers' is not known to be defined.
```

`elfeed-tube-utils.el` with checkdoc 0.6.2:
```
elfeed-tube-utils.el:271: All variables and subroutines might as well have a documentation string
elfeed-tube-utils.el:354: All variables and subroutines might as well have a documentation string
elfeed-tube-utils.el:362: All variables and subroutines might as well have a documentation string
elfeed-tube-utils.el:370: All variables and subroutines might as well have a documentation string
elfeed-tube-utils.el:431: All variables and subroutines might as well have a documentation string
```

`elfeed-tube-utils.el` with package-lint 20220607.738:
```
1 issue found:
62:19: error: You should depend on (emacs "28.1") if you need `ensure-list'.
```

`elfeed-tube-utils.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- elfeed-tube-utils.el#L300: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- elfeed-tube-utils.el#L301: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- elfeed-tube-utils.el#L302: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
```

`elfeed-tube.el` with byte-compile using Emacs 28.1:
```
elfeed-tube.el:521:35: Warning: reference to free variable
    `elfeed-tube--save-state-map'
In elfeed-tube--caption-echo:
elfeed-tube.el:699:15: Warning: reference to free variable
    `elfeed-tube--captions-echo-message'
```

`elfeed-tube.el` with checkdoc 0.6.2:
```
elfeed-tube.el:244: All interactive functions should have documentation
elfeed-tube.el:277: All variables and subroutines might as well have a documentation string
elfeed-tube.el:280: All variables and subroutines might as well have a documentation string
elfeed-tube.el:287: All variables and subroutines might as well have a documentation string
elfeed-tube.el:291: All variables and subroutines might as well have a documentation string
elfeed-tube.el:302: All variables and subroutines might as well have a documentation string
elfeed-tube.el:306: All variables and subroutines might as well have a documentation string
elfeed-tube.el:312: All variables and subroutines might as well have a documentation string
elfeed-tube.el:318: All variables and subroutines might as well have a documentation string
elfeed-tube.el:322: All variables and subroutines might as well have a documentation string
elfeed-tube.el:325: All variables and subroutines might as well have a documentation string
elfeed-tube.el:329: All variables and subroutines might as well have a documentation string
elfeed-tube.el:339: All variables and subroutines might as well have a documentation string
elfeed-tube.el:348: All variables and subroutines might as well have a documentation string
elfeed-tube.el:402: All variables and subroutines might as well have a documentation string
elfeed-tube.el:420: All variables and subroutines might as well have a documentation string
elfeed-tube.el:442: All variables and subroutines might as well have a documentation string
elfeed-tube.el:458: All variables and subroutines might as well have a documentation string
elfeed-tube.el:693: All variables and subroutines might as well have a documentation string
elfeed-tube.el:718: All variables and subroutines might as well have a documentation string
elfeed-tube.el:726: All variables and subroutines might as well have a documentation string
elfeed-tube.el:779: All variables and subroutines might as well have a documentation string
elfeed-tube.el:900: All variables and subroutines might as well have a documentation string
```

`elfeed-tube.el` with package-lint 20220607.738:
```
5 issues found:
450:9: error: You should depend on (emacs "28.1") if you need `replace-string-in-region'.
459:5: error: You should depend on (emacs "28.1") if you need `string-replace'.
634:38: error: You should depend on (emacs "28.1") if you need `string-replace'.
1127:26: error: You should depend on (emacs "28.1") if you need `ensure-list'.
1134:22: error: You should depend on (emacs "28.1") if you need `ensure-list'.
```

Package and license:
- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee

karthink commented:
I think I addressed all of the above in the latest commit.

- [x] Checkdoc warnings
- [x] byte-compile warnings
- [x] package-lint warnings
- [x] melpazoid warnings
- [x] License

Incidentally, `package-lint` did not show any of these errors when I ran it prior to submission here, or even today:

``` text
`elfeed-tube.el` with package-lint 20220607.738:
5 issues found:
450:9: error: You should depend on (emacs "28.1") if you need `replace-string-in-region'.
459:5: error: You should depend on (emacs "28.1") if you need `string-replace'.
634:38: error: You should depend on (emacs "28.1") if you need `string-replace'.
1127:26: error: You should depend on (emacs "28.1") if you need `ensure-list'.
1134:22: error: You should depend on (emacs "28.1") if you need `ensure-list'.

`elfeed-tube-utils.el` with package-lint 20220607.738:
1 issue found:
62:19: error: You should depend on (emacs "28.1") if you need `ensure-list'.
```

So I had to work off of your report. I got `No issues found.` instead. Any idea why this might be? I'm running the MELPA version fetched with straight.el, and it's been updated.

My apologies for missing the byte-compile errors. I was always byte-compiling the main file before `elfeed-tube-utils.el`, so everything it needed was already defined.

> I realize it's pedantic and probably more than I would personally normally do, but at the moment it's on the checklist and asked of all authors equally.

This is an argument for relaxing the requirement. For example, the docstring I added here is completely useless:

``` emacs-lisp
(defsubst elfeed-tube--thumbnail-html (thumb)
  "HTML for inserting THUMB."
  (when (and (elfeed-tube-include-p 'thumbnail) thumb)
    (concat "<br><img src=\"" thumb "\"></a><br><br>")))
```

Adding docstrings to two line `defsubst`s makes the code _harder_ to read. It's not possible to communicate intent via docstrings for smaller or internal functions either. Isn't requiring documentation for all interactive commands enough?

riscy commented:
> I got `No issues found.` instead. Any idea why this might be?

I'm not sure!  You could double-check the version you're using with `(pkg-info-format-version (pkg-info-package-version "package-lint"))` and compare to the version I'm using.  I also run my check through a [script](https://github.com/riscy/melpazoid) so the results I'm getting should be reproducible (you can even wire it into your GitHub actions).

> This is an argument for relaxing the requirement.

I have some ideas about how we might relax the checkdoc requirements on the backburner for a while (some of them are clearly more important than others).  But I haven't gotten around to formalizing them or raising an issue yet.  If there isn't already such an issue in the list of MELPA issues, you could raise one there as a starting point.

riscy commented:
Thanks for adding the license - I see it is "The Unlicense" but your files all have GPL boilerplate, so you should make a decision between those two.

Minor:
- With regard to your summary line, I guess YouTube is branded in camelcase, rather than as `Youtube` (which is in your summary line)

karthink commented:
``` emacs-lisp
(pkg-info-format-version (pkg-info-package-version "package-lint"))
```

``` text
"20220607.738"
```

This looks up to date, so I'm not sure. (This is relevant for `elfeed-tube-mpv`, another package I've placed in the MELPA queue.)

Thanks for the script, this looks like it will make my life much easier.

I've removed the GPL3 boilerplate from the files and changed Youtube -> YouTube in the summary.

riscy commented:
I'm not sure why I didn't catch this the first time around, but the individual files should have licenses too:

- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to elfeed-tube-utils.el
- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to elfeed-tube.el

karthink commented:
> I'm not sure why I didn't catch this the first time around

Because the individual files had the GPL3 boilerplate in them at that point!

I've added the following line:

SPDX-License-Identifier: UNLICENSE

to all files in the repo.

riscy commented:
Ha, I'd forgotten. :)

I think if you check the [SPDX list](https://spdx.org/licenses/) it should be written in titlecase (e.g. `Unlicense`) instead of `UNLICENSE` (the benefit of an exact match is an exact match to a formal/standardized list) but I won't block on that.  Merging!
